crates/engine/src/ext.rs

```diff
@@ -0,0 +1,430 @@
+    }
+
+    /// Conduct the BLAKE-2 128-bit hash and place the result into `output`.
+    pub fn hash_blake2_128(input: &[u8], output: &mut [u8; 16]) {
+        super::hashing::blake2b_128(input, output);
+    }
+
+    /// Conduct the SHA-2 256-bit hash and place the result into `output`.
+    pub fn hash_sha2_256(input: &[u8], output: &mut [u8; 32]) {
+        super::hashing::sha2_256(input, output);
+    }
+
+    /// Conduct the KECCAK 256-bit hash and place the result into `output`.
+    pub fn hash_keccak_256(input: &[u8], output: &mut [u8; 32]) {
+        super::hashing::keccak_256(input, output);
+    }
+
+    pub fn now(&self, _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `now`");
+    }
+
+    pub fn block_number(&self, _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `block_number`");
+    }
+
+    pub fn gas_left(&self, _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `gas_left`");
+    }
+
+    pub fn rent_allowance(&self, _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `rent_allowance`");
+    }
+
+    pub fn minimum_balance(&self, _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `minimum_balance`");
+    }
+
+    pub fn tombstone_deposit(&self, _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `tombstone_deposit`");
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn instantiate(
+        &mut self,
+        _code_hash: &[u8],
+        _gas_limit: u64,
+        _endowment: &[u8],
+        _input: &[u8],
+        _out_address: &mut &mut [u8],
+        _out_return_value: &mut &mut [u8],
+        _salt: &[u8],
+    ) -> Result {
+        unimplemented!("off-chain environment does not yet support `instantiate`");
+    }
+
+    pub fn call(
+        &mut self,
+        _callee: &[u8],
+        _gas_limit: u64,
+        _value: &[u8],
+        _input: &[u8],
+        _output: &mut &mut [u8],
+    ) -> Result {
+        unimplemented!("off-chain environment does not yet support `call`");
+    }
+
+    pub fn weight_to_fee(&self, _gas: u64, _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `weight_to_fee`");
+    }
+
+    pub fn set_rent_allowance(&mut self, _value: &[u8]) {
+        unimplemented!("off-chain environment does not yet support `set_rent_allowance`");
+    }
+
+    pub fn random(&self, _subject: &[u8], _output: &mut &mut [u8]) {
+        unimplemented!("off-chain environment does not yet support `random`");
+    }
+
+    pub fn call_chain_extension(
+        &mut self,
+        _func_id: u32,
+        _input: &[u8],
+        _output: &mut &mut [u8],
+    ) -> u32 {
+        unimplemented!(
+            "off-chain environment does not yet support `call_chain_extension`"
+        );
+    }
```

Comment:
Could you please create small issues enclosing follow-ups to this PR to add support for the various still missing pieces such as calling chain extension, `random`, and others?

Reply:
Done.